src/stream/src/executor/backfill/arrangement_backfill.rs

```diff
@@ -0,0 +1,516 @@
+                    }
+
+                    // We will update current_pos at least once,
+                    // since snapshot read has to be non-empty,
+                    // Or snapshot was empty and we construct a placeholder state.
+                    debug_assert_ne!(current_pos, None);
+
+                    persist_state(
+                        barrier.epoch,
+                        &mut self.state_table,
+                        true,
+                        &current_pos,
+                        &mut old_state,
+                        &mut current_state,
+                    )
+                    .await?;
+                    self.progress.finish(barrier.epoch.curr);
+                    yield msg;
+                    break;
+                }
+                yield msg;
+            }
+        }
+
+        // After progress finished + state persisted,
+        // we can forward messages directly to the downstream,
+        // as backfill is finished.
+        #[for_await]
+        for msg in upstream {
+            if let Some(msg) = mapping_message(msg?, &self.output_indices) {
+                if let Message::Barrier(barrier) = &msg {
+                    self.state_table.commit_no_data_expected(barrier.epoch);
+                }
+                yield msg;
+            }
+        }
+    }
+
+    /// Read snapshot per vnode.
+    /// These streams should be sorted in storage layer.
+    /// 1. Get row iterator / vnode.
+    /// 2. Merge it with futures unordered.
+    /// 3. Change it into a chunk iterator with `iter_chunks`.
+    /// This means it should fetch a row from each iterator to form a chunk.
+    /// Within each vnode, rows will be ordered by pk.
+    #[try_stream(ok = Option<StreamChunk>, error = StreamExecutorError)]
+    async fn snapshot_read_per_vnode<'a>(
+        schema: Arc<Schema>,
+        upstream_table: &'a ReplicatedStateTable<S>,
+        current_pos_map: &'a CurrentPosMap,
+    ) {
+        let mut streams = Vec::with_capacity(upstream_table.vnodes().len());
+        for vnode in upstream_table.vnodes().iter_vnodes() {
+            let current_pos = current_pos_map.get(&vnode);
+            let range_bounds = compute_bounds(upstream_table.pk_indices(), current_pos.cloned());
+            if range_bounds.is_none() {
+                continue;
+                // TODO: Highlight this change to reviewers.
+                // yield None;
+                // return Ok(());
+            }
+            let range_bounds = range_bounds.unwrap();
+            let vnode_iter = upstream_table
+                .iter_with_pk_range(&range_bounds, vnode, Default::default())
+                .await?;
+            streams.push(Box::pin(vnode_iter));
+        }
+        // TODO: Highlight this change to patrick vs `FuturesUnordered`, since `FuturesUnordered`
+        // merges `Futures` -> `Stream`, whereas `select_all` merges `Streams` -> `Stream`.
+        let iter = select_all(streams);
```

Comment:
Fyi I'm using `select_all` here, rather than `FuturesUnordered` as suggested, since `FuturesUnordered` merges `Futures` -> `Stream`, whereas `select_all` merges `Streams` -> `Stream`. `select_all` uses `FuturesUnordered` under the hood anyway.

Reply:
cc @hzxa21